mmedit/models/restorers/dic.py

```diff
@@ -0,0 +1,197 @@
+                log_vars, num_samples, results (lq, gt and pred).
+
+        """
+        # data
+        lq = data_batch['lq']
+        gt = data_batch['gt']
+        gt_heatmap = data_batch['heatmap']
+
+        # generate
+        sr_list, heatmap_list = self.generator.forward(lq)
+
+        # loss
+        losses = OrderedDict()
+
+        loss_pix = 0.0
+        loss_align = 0.0
+        for step, (sr, heatmap) in enumerate(zip(sr_list, heatmap_list)):
+            losses[f'loss_pixel_v{step}'] = self.pixel_loss(sr, gt)
+            loss_pix += losses[f'loss_pixel_v{step}']
+            losses[f'loss_align_v{step}'] = self.pixel_loss(
+                heatmap, gt_heatmap)
+            loss_align += losses[f'loss_align_v{step}']
+
+        # parse loss
+        loss, log_vars = self.parse_losses(losses)
+
+        # optimize
+        optimizer['generator'].zero_grad()
+        loss.backward()
+        optimizer['generator'].step()
+
+        log_vars.pop('loss')  # remove the unnecessary 'loss'
+        outputs = dict(
+            log_vars=log_vars,
+            num_samples=len(gt.data),
+            results=dict(lq=lq.cpu(), gt=gt.cpu(), output=sr_list[-1].cpu()))
+
+        return outputs
+
+    def forward_test(self,
+                     lq,
+                     gt=None,
+                     meta=None,
+                     save_image=False,
+                     save_path=None,
+                     iteration=None):
+        """Testing forward function.
+
+        Args:
+            lq (Tensor): LQ image.
+            gt (Tensor): GT image.
+            meta (list[dict]): Meta data, such as path of GT file.
+                Default: None.
+            save_image (bool): Whether to save image. Default: False.
+            save_path (str): Path to save image. Default: None.
+            iteration (int): Iteration for the saving image name.
+                Default: None.
+
+        Returns:
+            dict: Output results, which contain either key(s)
+                1. 'eval_result'.
+                2. 'lq', 'pred'.
+                3. 'lq', 'pred', 'gt'.
+        """
+
+        # generator
+        with torch.no_grad():
+            sr_list, _ = self.generator.forward(lq)
+            pred = sr_list[-1]
+            pred = self.img_denormalize(pred)
+
+            if gt is not None:
+                gt = self.img_denormalize(gt)
+
+        if self.test_cfg is not None and self.test_cfg.get('metrics', None):
+            assert gt is not None, (
+                'evaluation with metrics must have gt images.')
+            results = dict(eval_result=self.evaluate(pred, gt))
+        else:
+            results = dict(lq=lq.cpu(), output=pred.cpu())
+            if gt is not None:
+                results['gt'] = gt.cpu()
+
+        # save image
+        if save_image:
+            if 'gt_path' in meta[0]:
+                the_path = meta[0]['gt_path']
+            else:
+                the_path = meta[0]['lq_path']
```

Comment:
What does `the_path` mean? Why use `the`?

Reply:
get `pred_path` from `gt_path` or `lq_path`

Reply:
How about directly using `pred_path`?